lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp

```diff
@@ -965,22 +965,33 @@ static PartialApplyInst *lookThroughForPartialApply(SILValue V) {
 /// recognize when the apply takes a block argument that has been initialized to
 /// a non-escaping closure.
 static void checkForViolationsInNoEscapeClosures(
-    const StorageMap &Accesses, FullApplySite FAS, AccessSummaryAnalysis *ASA,
+    const StorageMap &Accesses, ApplySite AS, AccessSummaryAnalysis *ASA,
     llvm::SmallVectorImpl<ConflictingAccess> &ConflictingAccesses) {
-
-  SILFunction *Callee = FAS.getCalleeFunction();
-  if (Callee && !Callee->empty()) {
+  SILFunction *Callee = AS.getCalleeFunction();
+  if (isa<FullApplySite>(AS) && Callee && !Callee->empty()) {
     // Check for violation with directly called closure
-    checkForViolationWithCall(Accesses, Callee, 0, FAS.getArguments(), ASA,
+    checkForViolationWithCall(Accesses, Callee, 0, AS.getArguments(), ASA,
                               ConflictingAccesses);
   }
 
   // Check for violation with closures passed as arguments
-  for (SILValue Argument : FAS.getArguments()) {
+  for (SILValue Argument : AS.getArguments()) {
+    auto ArgumentFnType = Argument->getType().getAs<SILFunctionType>();
+    if (!ArgumentFnType)
+      continue;
+
+    if (!ArgumentFnType->isNoEscape())
+      continue;
+
     auto *PAI = lookThroughForPartialApply(Argument);
     if (!PAI)
       continue;
 
+    // Recursively check any arguments to the partial apply that are
+    // themselves no escape closures. This looks through reabstraction thunks.
+    checkForViolationsInNoEscapeClosures(Accesses, PAI, ASA,
+                                         ConflictingAccesses);
+
```

Comment:
It seems to me that `checkForViolationWithCall()` will be called twice for each closure passed to a call. Once here when we recurse, and again below.

Reply:
I've added an explicit check for whether the call is to a reabstraction thunk to prevent calling checkForViolationWithCall() on reabstraction thunks and instead perform the recursive case only.